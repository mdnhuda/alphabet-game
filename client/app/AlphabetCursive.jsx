import React from 'react';
import PropTypes from 'prop-types';
import {flattenCurve, scaleCurveCoOrdinates} from "./AlphabetUtils";
import LineOrQuadraticCurve from './LineOrQuadraticCurve';
import {Shape} from 'react-konva';

import {CurveShape} from "./ReactPropTypeShapes";

import {Stage, Layer, Line} from 'react-konva';

class AlphabetCursive extends React.Component {
    constructor(props) {
        super(props);

        console.log("in constructor");
        this.state = {
            ...props,
            stageWidth: props.origWidth
        };
    }

    componentDidMount() {
        console.log("in componentDidMount");
        this.checkSize();
        // here we should add listener for "container" resize
        // take a look here https://developers.google.com/web/updates/2016/10/resizeobserver
        // for simplicity I will just listen window resize
        window.addEventListener("resize", this.checkSize);
    }

    componentWillUnmount() {
        console.log("in componentWillUnmount");
        window.removeEventListener("resize", this.checkSize);
    }

    checkSize = () => {
        const width = this.container ? this.container.offsetWidth : this.state.origWidth;
        this.setState({
            stageWidth: width
        });
    };

    render() {
        console.log("in render");
        const {origCurves, origWidth, origHeight, stageWidth} = this.state;
        const strokeWidth = stageWidth / 50;
        const scale = stageWidth / origWidth;
        const scaledCurves = origCurves.map(curve => scaleCurveCoOrdinates(curve, {x: scale, y: 1}));
        return (
            <div style={{width: "100%", border: "1px solid grey"}} ref={node => {this.container = node;}}>
                <Stage width={stageWidth} height={origHeight}>
                    <Layer>
                        {
                            scaledCurves.map((curve, idx) => {
                                return <LineOrQuadraticCurve
                                    key={idx}
                                    points={curve}
                                    stroke='red'
                                    strokeWidth={strokeWidth}
                                />
                            })
                        }
                    </Layer>
                </Stage>
            </div>
        );
    }
}

AlphabetCursive.propTypes = {
    origCurves: PropTypes.arrayOf(CurveShape),
    origHeight: PropTypes.number,
    origWidth: PropTypes.number
};

export default AlphabetCursive;