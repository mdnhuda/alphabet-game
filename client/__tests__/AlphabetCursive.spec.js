import React from 'react';
import AlphabetCursive from '../app/AlphabetCursive';
import {shallow} from 'enzyme';

describe('AlphabetCursive', () => {
    beforeEach(() => {
        window = {
            ...window,
            innerHeight: 400,
            addEventListener: () => {},
            removeEventListener: () => {}
        }
    });

    it('with curves', () => {
        const component = shallow(
            <AlphabetCursive
                origCurves={[[{"x": 10, "y": 10}, {"x": 10, "y": 20}, {"x": 15, "y": 15}], [{"x": 20, "y": 40}, {"x": 20, "y": 60}]]}
                origWidth={200}
                origHeight={200}
            />
        );
        const instance = component.instance;
        instance.refs = {
            container: {
                offsetWidth: 500
            }
        };
        expect(component).toMatchSnapshot();
    });
});
