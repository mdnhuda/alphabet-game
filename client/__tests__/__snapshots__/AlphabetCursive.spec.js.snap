// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlphabetCursive with curves 1`] = `
<div
  style={
    Object {
      "border": "1px solid grey",
      "width": "100%",
    }
  }
>
  <Stage
    height={200}
    width={200}
  >
    <Layer>
      <LineOrQuadraticCurve
        key="0"
        points={
          Array [
            Object {
              "x": 10,
              "y": 10,
            },
            Object {
              "x": 10,
              "y": 20,
            },
            Object {
              "x": 15,
              "y": 15,
            },
          ]
        }
        stroke="red"
        strokeWidth={4}
      />
      <LineOrQuadraticCurve
        key="1"
        points={
          Array [
            Object {
              "x": 20,
              "y": 40,
            },
            Object {
              "x": 20,
              "y": 60,
            },
          ]
        }
        stroke="red"
        strokeWidth={4}
      />
    </Layer>
  </Stage>
</div>
`;
